### PR TITLE
Editor: Improve render performance (scrolling and typing)

### LINKED ISSUE #23427: Editor: Interaction triggering drastic browser repaint
**Describe the bug**

Screen Shot 2020-06-24 at 1 54 59 PM

I've noticed some lagginess with my local development. At first, I thought it may have been my computer. I did some testing, and it looks like there's some serious repainting happening in the editor - specifically from typing and scrolling.

I've recorded a video demo:
https://d.pr/v/NeLmSy

Every letter I type or every pixel I scroll triggers the entire page to redraw.

This can be tested by...

* Opening Gutenberg with Chrome
* Opening Chrome Developer Tools
* Press `Esc` to show extra panel
* Click on "Rendering"
* Toggle "Paint Flashing"

**Editor version (please complete the following information):**
- Latest Gutenberg `master`

**Desktop (please complete the following information):**
 - Mac OS
- Chrome, Version 83.0.4103.106

### PR DESCRIPTION
**Pull request by ItsJonQ**
85607658-520dc980-b622-11ea-8a7c-0d80a6ef5e1c

This update improves the rendering performance of the editor, especially for user interactions such as typing and scrolling.

This was achieved using a combination of CSS and JS updates that leverage GPU rendering where appropriate and optimize layout calculations.

Here are the results:

(Recorded on Chrome with the [paint rendering debug tool](https://d.pr/v/uAhLyG) enabled).
A green flash indicates a browser repaint.

**Before**
https://d.pr/v/g3eX8A

Notice that every interaction causes the entire editor to flash. It's bad enough when scrolling is involved. It's particularly concerning when the user types anything (anywhere, including the sidebar).

**After**
https://d.pr/v/I3uOyb

Wayyyy less repainting! Notice when typing, the repaint is contained in the typing area (e.g. RichText or `input`)

## How has this been tested?

This has been tested in local Gutenberg across Chrome, Firefox, and Safari.

* Run `npm install`
* Run `npm run dev`
* Use the editor (add content, etc...)
* Ensure that things render correctly

Certain things to pay attention to. Anything involving popovers (e.g. Toolbar, color picker).

For visual repaint testing...

* Enable Chrome's [paint rendering debug tool](https://d.pr/v/uAhLyG) (optional)
* Use the editor
* Analyze when/where things are repainting

## Types of changes

* Moved the `<Popover.Slot />` for BlockToolbar
* Added CSS layering with `translate3d(0, 0, 0)` for various elements

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [n/a] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [n/a] I've updated all React Native files affected by any refactorings/renamings in this PR. 

Resolves: https://github.com/WordPress/gutenberg/issues/23427

### CODE REVIEW

File: packages/components/src/toolbar/stories/index.js [hunk 1]
```diff
@@ -1,182 +0,0 @@
-/**
- * WordPress dependencies
- */
-import {
```
Comment (youknowriad) [inline packages/components/src/toolbar/stories/index.js:4]: Was this removal intended?

File: packages/edit-post/src/components/layout/index.js [hunk 2]
```diff
@@ -266,6 +266,7 @@ function Layout() {
 				<KeyboardShortcutHelpModal />
 				<WelcomeGuide />
 				<Popover.Slot />
+				<Popover.Slot name="block-toolbar" />
```
Comment (youknowriad) [inline packages/edit-post/src/components/layout/index.js:269]: I believe this slot was not rendered her for a valid reason. Otherwise, it just duplicated the slot that is above it. cc @ellatrix

**Review by youknowriad — COMMENTED** [review]
I love this 💯 Thanks a lot, this is very cool.

**Comment by ZebulanStanphill** [comment]
Instead of the clever, but somewhat hacky method of using `transform` to trigger GPU optimizations, why not try using the [`contain`](https://developer.mozilla.org/en-US/docs/Web/CSS/contain) property instead?

**Comment by ItsJonQ** [comment]
@ZebulanStanphill Oh interesting! I didn't know about `contain`. It looks like it's not support on **Safari** and **IE**. Bummer :(

File: packages/components/src/toolbar/stories/index.js (hunk 1 above)
Comment (ItsJonQ) [reply packages/components/src/toolbar/stories/index.js:4]: 🤦‍♀️ Ah yes... During dev, something happened with my `git` and I noticed `toolbar` may have been affected? But I wasn't able to spot. Thank you for catching this! I'll revert this now

File: packages/edit-post/src/components/layout/index.js (hunk 2 above)
Comment (ItsJonQ) [reply packages/edit-post/src/components/layout/index.js:269]: @youknowriad Hmm! Interesting.. I wasn't aware of that !

**Comment by ItsJonQ** [comment]
Hmm.. Not sure how to debug the failing E2E tests.
Locally, the tests fail for me on `master` as well 🙈

**Comment by ntsekouras** [comment]
Hey @ItsJonQ -- great additions here!

>Locally, the tests fail for me on master as well

Have you tried commenting `observeConsoleLogging()` for running e2e locally? I tried some of them locally on master and they pass ( with the above commented ).

I'll take a look now on your branch to see if I can help :)

**Comment by ItsJonQ** [comment]
@ntsekouras Thanks for helping!

I'm not too sure how to resolve this. The **test** passes for me (locally) in interactive mode.

```
npm run test-e2e:watch -- packages/e2e-tests/specs/editor/various/multi-block-selection.test.js --puppeteer-interactive --puppeteer-slowmo=100
```

(I run it with `.only`, so it targets the failing test)

```
	it.only( 'should clear selection when clicking next to blocks', async () => { ... } );
```

File: packages/edit-post/src/components/layout/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/edit-post/src/components/layout/index.js:269]: This needs to be rendered in the content scroll container, because when you scroll while the pointer is over the toolbar, the content still needs to scroll.

Comment (ellatrix) [reply packages/edit-post/src/components/layout/index.js:269]: I should have commented here 😅
There might be other reasons I'm forgetting.

**Comment by ellatrix** [comment]
What are the performance results compared to `master`?

**Comment by ItsJonQ** [comment]
> What are the performance results compared to master?

@ellatrix Excellent question! I just ran some Performance + FPS tests against master.

### Typing

I think typing is where we'd see/feel noticeable improvements.

In the following performance tests, I typed "Typing into the editor" into a Paragraph block.

### Before

typing-before

### After

typing-after

(Higher overall FPS levels)

The performance tool makes things laggier (in general). However, it felt smoother typing with the new render improvements.

---

### Scrolling

Not a huge change. But a bit!

#### Before

scrolling-before

#### After

scrolling-after

Looks like lower CPU spikes during start/stops.

#### Scrolling FPS

#### Before

scrolling-before

(Up to approx. 50-51FPS on heavy scrolling)


#### After

scrolling-after

(Up to approx. 54-55FPS on heavy scrolling)

**Comment by ItsJonQ** [comment]
Updates. After a bunch of poking and testing... it looks like the E2E tests aren't happy with the `translate3d(0, 0, 0)` change applied to:

```
.interface-interface-skeleton__content,
.interface-interface-skeleton__left-sidebar,
.interface-interface-skeleton__sidebar
```

Maybe it's something up with Puppeteer? Maybe not.
It looks like the Android test is unhappy as well.

Unfortunately, that is important piece for this performance enhancement, which attempts to tame repaint/layout calc.

Maybe we should give up on this for now?

**Comment by ZebulanStanphill** [comment]
> @ZebulanStanphill Oh interesting! I didn't know about `contain`. It looks like it's not support on **Safari** and **IE**. Bummer :(

I think it's still worth trying out as a progressive enhancement for the browsers that do support it, regardless of whether we also proceed with this PR or not.

**Comment by ItsJonQ** [comment]
> I think it's still worth trying out as a progressive enhancement for the browsers that do support it, regardless of whether we also proceed with this PR or not.

@ZebulanStanphill That's fair. What's strange is...  I experimented with `contain`, but it didn't seem to make a difference (repaint/performance wise). `translate3d` did though.

**Comment by ItsJonQ** [comment]
Closing this for now. The initial research was helpful in identifying the problem areas.
